go/vt/wrangler/stream_migrater.go

```diff
@@ -0,0 +1,604 @@
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	var oneSet []*vrStream
+	allErrors := &concurrency.AllErrorRecorder{}
+	for _, tabletStreams := range stoppedStreams {
+		if oneSet == nil {
+			oneSet = tabletStreams
+		}
+		for _, vrs := range tabletStreams {
+			key := fmt.Sprintf("%s:%s", vrs.bls.Keyspace, vrs.bls.Shard)
+			pos := stopPositions[key]
+			if !vrs.pos.Equal(pos) {
+				allErrors.RecordError(fmt.Errorf("%s: stream %d position: %s does not match %s", key, vrs.id, mysql.EncodePosition(vrs.pos), mysql.EncodePosition(pos)))
+			}
+		}
+	}
+	return oneSet, allErrors.AggrError(vterrors.Aggregate)
+}
+
+func (sm *streamMigrater) migrateStreams(ctx context.Context, tabletStreams []*vrStream) ([]string, error) {
+	if sm.mi.migrationType == binlogdatapb.MigrationType_TABLES {
+		return nil, nil
+	}
+
+	// Delete any previous stray workflows that might have been left-over
+	// due to a failed migration.
+	if err := sm.deleteTargetStreams(ctx, tabletStreams); err != nil {
+		return nil, err
+	}
+
+	tmpl, err := sm.templatize(ctx, tabletStreams)
+	if err != nil {
+		return nil, err
+	}
+	workflows := tabletStreamWorkflows(tmpl)
+	if err := sm.createTargetStreams(ctx, tmpl); err != nil {
+		return nil, err
+	}
+	return workflows, nil
+}
+
+const (
+	unknown = iota
+	sharded
+	reference
+)
+
+func (sm *streamMigrater) templatize(ctx context.Context, tabletStreams []*vrStream) ([]*vrStream, error) {
+	tabletStreams = copyTabletStreams(tabletStreams)
+	var shardedStreams []*vrStream
+	for _, vrs := range tabletStreams {
+		streamType := unknown
+		for _, rule := range vrs.bls.Filter.Rules {
+			typ, err := sm.templatizeRule(ctx, rule)
+			if err != nil {
+				return nil, err
+			}
+			switch typ {
+			case sharded:
+				if streamType == reference {
+					return nil, fmt.Errorf("cannot migrate streams with a mix of reference and sharded tables: %v", vrs.bls)
+				}
+				streamType = sharded
+			case reference:
+				if streamType == sharded {
+					return nil, fmt.Errorf("cannot migrate streams with a mix of reference and sharded tables: %v", vrs.bls)
+				}
+				streamType = reference
+			}
+		}
+		if streamType == sharded {
+			shardedStreams = append(shardedStreams, vrs)
+		}
+	}
+	return shardedStreams, nil
+}
+
+func (sm *streamMigrater) templatizeRule(ctx context.Context, rule *binlogdatapb.Rule) (int, error) {
+	switch {
+	case rule.Filter == "":
+		return reference, nil
+	case key.IsKeyRange(rule.Filter):
+		rule.Filter = "{{.}}"
+		return sharded, nil
+	case rule.Filter == vreplication.ExcludeStr:
+		return unknown, nil
+	default:
+		templatized, err := sm.checkShardedQuery(ctx, rule.Filter)
+		if err != nil {
+			return unknown, err
+		}
+		if templatized != "" {
+			rule.Filter = templatized
+			return sharded, nil
+		}
+		return reference, nil
+	}
+}
+
+func (sm *streamMigrater) checkShardedQuery(ctx context.Context, query string) (string, error) {
+	statement, err := sqlparser.Parse(query)
```

Comment:
Can you add some comments to this function describing what we are checking for?

Reply:
Renamed the function to templatizeQuery and added comments.